teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  or vice versa?
+    * The TaskRun (or Pipeline Task) would fail
+* What if the Task writes to both the string file and the json file for the same param?
+    * The TaskRun (or Pipeline Task) would fail
+
+### Risks and Mitigations
+
+* Changing the type of Task Result values from string to being ArrayOrString might be backwards incompatible
+    * Mitigation: perform a manual test to make sure this isn't backwards incompatible (reconcile a TaskRun with a string
+      result both before and after the change - e.g. create a TaskRun before the change, then restart the controller so
+      it is re-reconciled; revisit this proposal if we can't find a way to make the change and be backward compatible)
+    * Note this will be a change for folks consuming the library, e.g. the CLI, as a result implementing this TEP
+      should include working with the CLI team to pull in the updates and make sure they are not negatively impacted
+
+## Design Details
+
+### Emitting array results
+
+We add support for writing json content to `/tekton/results/resultName.json`, supporting strings and arrays of strings.
+
+For example, say we want the value of the Task foo's result `animals` to be "cat", "dog" and "squirrel":
+
+1. Write the following content to `$(results.animals.jsonPath)`:
+
+   ```json
+   ["cat", "dog", "squirrel"]
+   ```
+
+2. This would be written to the pod termination message as escaped json, for example (with a string example included as well):
+
+   ```
+   message: '[{"key":"animals","value":"[\"cat\", \"dog\", \"squirrel\"]","type":"TaskRunResult"},{"key":"someString","value":"aStringValue","type":"TaskRunResult"}]'
+   ```
+
+3. We would use the same
+   [ArrayOrString type](https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/pkg/apis/pipeline/v1beta1/param_types.go#L89)
+   (which in TEP-0075 could be expanded to support dictionaries as well) for
+   task results, e.g. for the above example, the TaskRun would contain:
+
+   ```yaml
+     taskResults:
+     - name: someString
+       value: aStringValue
+     - name: animals
+       value:
+       - cat
+       - dog
+       - squirrel
+   ```
+
+### Indexing into arrays with variable replacement
+
+In addition to supporting [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+style [star notation](https://github.com/tektoncd/pipeline/issues/2041), we add support for
+[the JSONPath subscript operator](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html#name-jsonpath-examples)
+for accessing "child" members of the array by index.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+...
+spec:
+  params:
+    - environments:
+        - 'staging'
+        - 'qa'
+        - 'prod'
+...
+  tasks:
+    - name: deploy
+      params:
+        - name: environment
+          value: '$(params.environments[0])'
```

Comment:
this is where I think dynamism could be super cool. What if we did:
```suggestion
  tasks:
    - name: deploy
      params:
        - name: environment
          value: '$(params.environments[])'
```

Having the non-indexed array there, similar to `jq` would produce a copy of task deploy for each valid in the input `params.environments`. Seems fairly straight-forward and a reasonable addition to the api.

Reply:
I disagree there, or at least in the scope of the TEP. This might be something that appear confusing to the users, and most importantly, I would like to make sure we really need this (and custom task are not enough) before commiting to that level of complexity.

Reply:
I can think of many use cases related to test sharding where this would be super useful. We also have many "build these X services and deploy them to an ephemeral cluster" uses cases. All involve a ton of copy+pasta of the same exact spec with different params

Reply:
Right, but as of today, this is what `task-loop` custom task is there for, isn't it ?

Reply:
@pierretasci  100% agree with the test sharding use case and i have a languishing todo list item about trying to write up a TEP about better supporting test sharding specifically (which ultimately would lead to needing looping support)

I also agree with @vdemeester though that I don't personally see that as being exaclty within the scope of this particular TEP to add looping support BUT I _do_ see array result support as a requirement for being able to support looping (and sharding) - and maybe that's what you're getting at @pierretasci (not that this TEP includes implementing looping, but that we need this TEP for looping and sharding is a great use case to list to justify adding array results)

Reply:
Well I was specifically advocating for supporting it directly in the Tekton API because I think that unlocks many more use-cases but I can also see the argument against.

Reply:
i am 💯  in favor of looping in the pipelines api at this point but i think we need to explore the use cases and alternatives a bit (and hopefully @vdemeester will find them convincing :D )

Reply:
https://github.com/tektoncd/community/pull/600 🎉 🎉 🎉